client/test/fund.go

```diff
@@ -0,0 +1,160 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"math/rand"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/client"
+	"perun.network/go-perun/wire"
+	"polycry.pt/poly-go/test"
+)
+
+// FundSetup represents the fund recovery test parameters.
+type FundSetup struct {
+	ChallengeDuration uint64
+	FridaInitBal      channel.Bal
+	FredInitBal       channel.Bal
+	BalanceDelta      channel.Bal
+}
+
+// TestFundRecovery performs a test of the fund recovery functionality for the
+// given setup.
+func TestFundRecovery(ctx context.Context, t *testing.T, params FundSetup, setup func(*rand.Rand) ([2]RoleSetup, channel.Asset)) { //nolint:revive // test.Test... stutters but this is OK in this special case.
+	rng := test.Prng(t)
+
+	t.Run("failing funder proposer", func(t *testing.T) {
+		roles, asset := setup(rng)
+		roles[0].Funder = FailingFunder{}
+
+		runFredFridaTest(ctx, t, rng, params, roles, asset)
+	})
+
+	t.Run("failing funder proposee", func(t *testing.T) {
+		roles, asset := setup(rng)
+		roles[1].Funder = FailingFunder{}
+
+		runFredFridaTest(ctx, t, rng, params, roles, asset)
+	})
+
+	t.Run("failing funder both sides", func(t *testing.T) {
+		roles, asset := setup(rng)
+		roles[0].Funder = FailingFunder{}
+		roles[1].Funder = FailingFunder{}
+
+		runFredFridaTest(ctx, t, rng, params, roles, asset)
+	})
+}
+
+//nolint:thelper // The linter thinks this is a helper function, but it is not.
+func runFredFridaTest(
+	ctx context.Context,
+	t *testing.T,
+	rng *rand.Rand,
+	params FundSetup,
+	setups [2]RoleSetup,
+	asset channel.Asset,
+) {
+	const (
+		fridaIdx = 0
+		fredIdx  = 1
+		numParts = 2
+	)
+	var (
+		challengeDuration = params.ChallengeDuration
+		fridaInitBal      = params.FridaInitBal
+		fredInitBal       = params.FredInitBal
+	)
+
+	clients := NewClients(t, rng, setups[:])
+	frida, fred := clients[fridaIdx], clients[fredIdx]
+	fridaWireAddr, fredWireAddr := frida.Identity.Address(), fred.Identity.Address()
+	fridaWalletAddr, fredWalletAddr := frida.WalletAddress, fred.WalletAddress
+
+	// Store client balances before running test.
+	balancesBefore := channel.Balances{
+		{
+			frida.BalanceReader.Balance(asset),
+			fred.BalanceReader.Balance(asset),
+		},
+	}
+
+	// The channel into which Fred's created ledger channel is sent into.
```

Comment:
```suggestion
	// The channel into which Fred's created ledger channel is sent.
```
I think the mistake even comes from me when I originally created the tests but I didn't see it until now.

Reply:
Revised the comment.